src/main/java/core/basesyntax/operations/PurchaseOperation.java

```diff
@@ -0,0 +1,29 @@
+package core.basesyntax.operations;
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.FruitTransaction;
+
+public class PurchaseOperation implements OperationHandler {
+    private final Storage storage;
+
+    public PurchaseOperation(Storage storage) {
+        this.storage = storage;
+    }
+
+    @Override
+    public void apply(FruitTransaction transaction) {
+        if (transaction == null) {
+            throw new IllegalArgumentException("Transaction cannot be null.");
+        }
+
+        int currentQuantity = storage.getQuantity(transaction.getFruit());
+        int newQuantity = currentQuantity - transaction.getQuantity();
+
+        if (newQuantity < 0) {
+            throw new RuntimeException("Insufficient stock for purchase of "
+                    + transaction.getFruit());
+        }
+
+        storage.updateFruit(transaction.getFruit(), newQuantity);

```

Comment:
Following the encapsulation principle, consider making `storage` private and provide interaction through methods rather than direct access.